File: packages/sdk-go/crypto_orders_test.go
package driftstack

import (
	"context"
	"encoding/json"
	"net/http"
	"testing"
)

func TestCryptoOrders_Quote(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/billing/crypto-checkout/quote" || r.Method != "POST" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"price_cents":999,"asset":"usdc"}`))
	})
	got, err := client.CryptoOrders.Quote(context.Background(), map[string]any{
		"product": "solo_manual",
	})
	if err != nil {
		t.Fatal(err)
	}
	if got["price_cents"].(float64) != 999 {
		t.Errorf("price_cents=%v", got["price_cents"])
	}
}

func TestCryptoOrders_CreateCheckout_ForwardsIdempotencyKey(t *testing.T) {
	t.Parallel()
	var sawKey string
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		sawKey = r.Header.Get("Idempotency-Key")
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"order_id":"ord_1"}`))
	})
	key := "abc-123"
	_, err := client.CryptoOrders.CreateCheckout(
		context.Background(),
		map[string]any{"product": "solo_manual"},
		&CreateCheckoutOptions{IdempotencyKey: &key},
	)
	if err != nil {
		t.Fatal(err)
	}
	if sawKey != "abc-123" {
		t.Errorf("Idempotency-Key header = %q, want %q", sawKey, "abc-123")
	}
}

func TestCryptoOrders_CreateCheckout_PreservesAuthHeader(t *testing.T) {
	// The headers map in requestOptions is merged AFTER Authorization /
	// User-Agent / Content-Type are set; a future refactor mustn't let
	// caller-supplied headers (e.g. Idempotency-Key) clobber the
	// bearer token by accident.
	t.Parallel()
	var sawAuth, sawKey, sawUA string
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		sawAuth = r.Header.Get("Authorization")
		sawKey = r.Header.Get("Idempotency-Key")
		sawUA = r.Header.Get("User-Agent")
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{}`))
	})
	key := "k-1"
	if _, err := client.CryptoOrders.CreateCheckout(
		context.Background(),
		map[string]any{"product": "solo_manual"},
		&CreateCheckoutOptions{IdempotencyKey: &key},
	); err != nil {
		t.Fatal(err)
	}
	if sawAuth != "Bearer ds_test_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" {
		t.Errorf("Authorization header = %q, want Bearer …", sawAuth)
	}
	if sawKey != "k-1" {
		t.Errorf("Idempotency-Key = %q", sawKey)
	}
	if sawUA == "" {
		t.Errorf("User-Agent missing")
	}
}

func TestCryptoOrders_CreateCheckout_NoKeyOmitsHeader(t *testing.T) {
	t.Parallel()
	var sawKey string
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		sawKey = r.Header.Get("Idempotency-Key")
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"order_id":"ord_1"}`))
	})
	if _, err := client.CryptoOrders.CreateCheckout(
		context.Background(),
		map[string]any{"product": "solo_manual"},
		nil,
	); err != nil {
		t.Fatal(err)
	}
	if sawKey != "" {
		t.Errorf("expected no Idempotency-Key header, got %q", sawKey)
	}
}

func TestCryptoOrders_List_PassesThroughFilters(t *testing.T) {
	t.Parallel()
	var sawQuery string
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		sawQuery = r.URL.RawQuery
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"orders":[],"next_cursor":null}`))
	})
	lim := 25
	status := "paid"
	cur := "c0"
	after := "2026-05-01T00:00:00Z"
	before := "2026-06-01T00:00:00Z"
	_, err := client.CryptoOrders.List(context.Background(), &ListCryptoOrdersOptions{
		Limit:         &lim,
		Status:        &status,
		Cursor:        &cur,
		CreatedAfter:  &after,
		CreatedBefore: &before,
	})
	if err != nil {
		t.Fatal(err)
	}
	want := []string{
		"limit=25",
		"status=paid",
		"cursor=c0",
		"created_after=2026-05-01T00%3A00%3A00Z",
		"created_before=2026-06-01T00%3A00%3A00Z",
	}
	for _, w := range want {
		if !contains(sawQuery, w) {
			t.Errorf("query %q missing %q", sawQuery, w)
		}
	}
}

func TestCryptoOrders_List_OmitsUnsetParams(t *testing.T) {
	t.Parallel()
	var sawQuery string
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		sawQuery = r.URL.RawQuery
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"orders":[]}`))
	})
	if _, err := client.CryptoOrders.List(context.Background(), nil); err != nil {
		t.Fatal(err)
	}
	if sawQuery != "" {
		t.Errorf("expected no query string, got %q", sawQuery)
	}
}

func TestCryptoOrders_Iterate_WalksCursorPages(t *testing.T) {
	t.Parallel()
	c1 := "c1"
	c2 := "c2"
	pages := []ListCryptoOrdersResponse{
		{Orders: []CryptoOrderEnvelope{{"id": "ord_1"}}, NextCursor: &c1},
		{Orders: []CryptoOrderEnvelope{{"id": "ord_2"}, {"id": "ord_3"}}, NextCursor: &c2},
		{Orders: []CryptoOrderEnvelope{{"id": "ord_4"}}, NextCursor: nil},
	}
	seenCursors := []string{}
	idx := 0
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		seenCursors = append(seenCursors, r.URL.Query().Get("cursor"))
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(pages[idx])
		idx++
	})
	var ids []string
	err := client.CryptoOrders.Iterate(context.Background(), nil, func(o CryptoOrderEnvelope) bool {
		ids = append(ids, o["id"].(string))
		return true
	})
	if err != nil {
		t.Fatal(err)
	}
	wantIDs := []string{"ord_1", "ord_2", "ord_3", "ord_4"}
	if len(ids) != len(wantIDs) {
		t.Fatalf("ids=%v, want=%v", ids, wantIDs)
	}
	for i, id := range ids {
		if id != wantIDs[i] {
			t.Errorf("ids[%d]=%q, want %q", i, id, wantIDs[i])
		}
	}
	wantCursors := []string{"", "c1", "c2"}
	if len(seenCursors) != len(wantCursors) {
		t.Fatalf("cursors=%v, want=%v", seenCursors, wantCursors)
	}
	for i, c := range seenCursors {
		if c != wantCursors[i] {
			t.Errorf("cursors[%d]=%q, want %q", i, c, wantCursors[i])
		}
	}
}

func TestCryptoOrders_Iterate_TreatsEmptyCursorAsTerminal(t *testing.T) {
	t.Parallel()
	empty := ""
	pages := []ListCryptoOrdersResponse{
		{Orders: []CryptoOrderEnvelope{{"id": "ord_1"}}, NextCursor: &empty},
	}
	idx := 0
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(pages[idx])
		idx++
	})
	var ids []string
	err := client.CryptoOrders.Iterate(context.Background(), nil, func(o CryptoOrderEnvelope) bool {
		ids = append(ids, o["id"].(string))
		return true
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(ids) != 1 || ids[0] != "ord_1" {
		t.Errorf("ids=%v, want [ord_1]", ids)
	}
	// An empty-string next_cursor must be treated as terminal (matching the
	// audit-log / profiles / profile-snapshots iterators). Otherwise the loop
	// would re-fetch with cursor="" and over-run the single-page slice.
	if idx != 1 {
		t.Errorf("fetched %d pages, want 1 (empty cursor should be terminal)", idx)
	}
}

func TestCryptoOrders_Iterate_StopsEarlyOnVisitFalse(t *testing.T) {
	t.Parallel()
	c1 := "c1"
	pages := []ListCryptoOrdersResponse{
		{Orders: []CryptoOrderEnvelope{{"id": "ord_1"}, {"id": "ord_2"}}, NextCursor: &c1},
		// Should not be fetched — visit returns false on ord_1.
		{Orders: []CryptoOrderEnvelope{{"id": "ord_3"}}, NextCursor: nil},
	}
	idx := 0
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(pages[idx])
		idx++
	})
	var ids []string
	err := client.CryptoOrders.Iterate(context.Background(), nil, func(o CryptoOrderEnvelope) bool {
		ids = append(ids, o["id"].(string))
		// Stop after the first order.
		return false
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(ids) != 1 || ids[0] != "ord_1" {
		t.Errorf("ids=%v, want [ord_1]", ids)
	}
	if idx != 1 {
		t.Errorf("fetched %d pages, want 1", idx)
	}
}

func TestCryptoOrders_Get_EncodesOrderID(t *testing.T) {
	t.Parallel()
	var sawPath string
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		// RequestURI preserves percent-encoding; Path is decoded.
		sawPath = r.RequestURI
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"id":"co/with/slash"}`))
	})
	if _, err := client.CryptoOrders.Get(context.Background(), "co/with/slash"); err != nil {
		t.Fatal(err)
	}
	// url.PathEscape encodes '/' to %2F.
	if sawPath != "/v1/billing/crypto-orders/co%2Fwith%2Fslash" {
		t.Errorf("path=%q, want encoded slashes", sawPath)
	}
}

func TestCryptoOrders_Receipt_Path(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/billing/crypto-orders/ord_1/receipt" {
			t.Errorf("path=%q", r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"order_id":"ord_1"}`))
	})
	out, err := client.CryptoOrders.Receipt(context.Background(), "ord_1")
	if err != nil {
		t.Fatal(err)
	}
	if out["order_id"] != "ord_1" {
		t.Errorf("order_id=%v", out["order_id"])
	}
}

func TestCryptoOrders_Cancel_Path(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/billing/crypto-orders/ord_1/cancel" || r.Method != "POST" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"id":"ord_1","status":"cancelled"}`))
	})
	out, err := client.CryptoOrders.Cancel(context.Background(), "ord_1")
	if err != nil {
		t.Fatal(err)
	}
	if out["status"] != "cancelled" {
		t.Errorf("status=%v", out["status"])
	}
}

func TestCryptoOrders_UpdateNote_PATCH(t *testing.T) {
	t.Parallel()
	var sawBody map[string]any
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "PATCH" {
			t.Errorf("method=%q, want PATCH", r.Method)
		}
		_ = json.NewDecoder(r.Body).Decode(&sawBody)
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"id":"ord_1","customer_note":"PO-1"}`))
	})
	if _, err := client.CryptoOrders.UpdateNote(
		context.Background(),
		"ord_1",
		map[string]any{"customer_note": "PO-1"},
	); err != nil {
		t.Fatal(err)
	}
	if sawBody["customer_note"] != "PO-1" {
		t.Errorf("body=%v", sawBody)
	}
}

// helper — `strings.Contains` would do, but pulling the import in just
// for tests bloats the file. Keep this tiny local one to match the
// pattern in client_test.go.
func contains(haystack, needle string) bool {
	for i := 0; i+len(needle) <= len(haystack); i++ {
		if haystack[i:i+len(needle)] == needle {
			return true
		}
	}
	return false
}

File: packages/sdk-typescript/tests/unit/crypto-orders.test.ts
// V-666.BB — CryptoOrdersResource unit tests.

import { describe, expect, it, vi } from 'vitest';
import { CryptoOrdersResource } from '../../src/resources/crypto-orders.js';
import type { HttpClient } from '../../src/http.js';

interface RequestOpts {
  method: string;
  path: string;
  body?: unknown;
  headers?: Record<string, string>;
  query?: Record<string, string | number | undefined>;
}

function harness(): {
  http: HttpClient;
  calls: RequestOpts[];
  setResponse: (response: unknown) => void;
} {
  const calls: RequestOpts[] = [];
  let nextResponse: unknown = {};
  const request = vi.fn((opts: RequestOpts) => {
    calls.push(opts);
    return Promise.resolve(nextResponse);
  });
  return {
    http: { request } as unknown as HttpClient,
    calls,
    setResponse: (r) => {
      nextResponse = r;
    },
  };
}

describe('CryptoOrdersResource', () => {
  it('quote POSTs /v1/billing/crypto-checkout/quote with body', async () => {
    const h = harness();
    h.setResponse({ product: 'solo_manual', price_cents: 2500 });
    const r = new CryptoOrdersResource(h.http);
    await r.quote({ product: 'solo_manual' });
    expect(h.calls[0]?.method).toBe('POST');
    expect(h.calls[0]?.path).toBe('/v1/billing/crypto-checkout/quote');
    expect(h.calls[0]?.body).toEqual({ product: 'solo_manual' });
  });

  it('createCheckout POSTs without an idempotency header when none supplied', async () => {
    const h = harness();
    h.setResponse({ order_id: 'ord_a' });
    const r = new CryptoOrdersResource(h.http);
    await r.createCheckout({ product: 'solo_manual', price_cents: 7900, price_currency: 'USD' });
    expect(h.calls[0]?.path).toBe('/v1/billing/crypto-checkout');
    expect(h.calls[0]?.headers).toBeUndefined();
  });

  it('createCheckout passes Idempotency-Key when supplied', async () => {
    const h = harness();
    h.setResponse({ order_id: 'ord_a' });
    const r = new CryptoOrdersResource(h.http);
    await r.createCheckout(
      { product: 'solo_manual', price_cents: 7900, price_currency: 'USD' },
      { idempotencyKey: 'k-123' },
    );
    expect(h.calls[0]?.headers).toEqual({ 'idempotency-key': 'k-123' });
  });

  it('list GETs the list endpoint', async () => {
    const h = harness();
    h.setResponse({ orders: [] });
    const r = new CryptoOrdersResource(h.http);
    await r.list();
    expect(h.calls[0]?.method).toBe('GET');
    expect(h.calls[0]?.path).toBe('/v1/billing/crypto-orders');
    expect(h.calls[0]?.query).toBeUndefined();
  });

  it('V-666.BU listAll walks pages until next_cursor is null', async () => {
    const calls: unknown[] = [];
    // Wire a custom request handler that emits two pages.
    const requestFn = vi.fn((opts: RequestOpts) => {
      calls.push(opts);
      if (opts.query?.cursor === undefined) {
        return Promise.resolve({
          orders: [{ order_id: 'ord_1' }, { order_id: 'ord_2' }],
          next_cursor: 'cur_x',
        });
      }
      return Promise.resolve({
        orders: [{ order_id: 'ord_3' }],
        next_cursor: null,
      });
    });
    const r = new CryptoOrdersResource({ request: requestFn } as unknown as HttpClient);
    const collected: string[] = [];
    for await (const o of r.listAll({ status: 'paid' })) {
      collected.push(o.order_id);
    }
    expect(collected).toEqual(['ord_1', 'ord_2', 'ord_3']);
    expect(calls.length).toBe(2);
    const first = calls[0] as RequestOpts;
    const second = calls[1] as RequestOpts;
    expect(first.query).toEqual({ status: 'paid' });
    expect(second.query).toEqual({ status: 'paid', cursor: 'cur_x' });
  });

  it('V-666.BU listAll handles a single page (next_cursor null) without a second fetch', async () => {
    let calls = 0;
    const requestFn = vi.fn(() => {
      calls++;
      return Promise.resolve({
        orders: [{ order_id: 'ord_only' }],
        next_cursor: null,
      });
    });
    const r = new CryptoOrdersResource({ request: requestFn } as unknown as HttpClient);
    const collected: string[] = [];
    for await (const o of r.listAll()) collected.push(o.order_id);
    expect(collected).toEqual(['ord_only']);
    expect(calls).toBe(1);
  });

  it('V-666.BU listAll yields nothing when the first page is empty', async () => {
    const requestFn = vi.fn(() => Promise.resolve({ orders: [], next_cursor: null }));
    const r = new CryptoOrdersResource({ request: requestFn } as unknown as HttpClient);
    const collected: string[] = [];
    for await (const o of r.listAll()) collected.push(o.order_id);
    expect(collected).toEqual([]);
  });

  it('V-666.BU listAll respects early break (no extra requests)', async () => {
    let calls = 0;
    const requestFn = vi.fn(() => {
      calls++;
      return Promise.resolve({
        orders: [
          { order_id: `ord_${calls.toString()}_a` },
          { order_id: `ord_${calls.toString()}_b` },
        ],
        next_cursor: 'cur_more',
      });
    });
    const r = new CryptoOrdersResource({ request: requestFn } as unknown as HttpClient);
    for await (const o of r.listAll()) {
      // Break after the very first yielded order.
      expect(o.order_id).toBe('ord_1_a');
      break;
    }
    expect(calls).toBe(1);
  });

  it('V-666.BU iterate alias walks pages identically to listAll (cross-SDK naming parity)', async () => {
    const calls: unknown[] = [];
    const requestFn = vi.fn((opts: RequestOpts) => {
      calls.push(opts);
      if (opts.query?.cursor === undefined) {
        return Promise.resolve({
          orders: [{ order_id: 'ord_1' }, { order_id: 'ord_2' }],
          next_cursor: 'cur_x',
        });
      }
      return Promise.resolve({ orders: [{ order_id: 'ord_3' }], next_cursor: null });
    });
    const r = new CryptoOrdersResource({ request: requestFn } as unknown as HttpClient);
    const collected: string[] = [];
    for await (const o of r.iterate({ status: 'paid' })) collected.push(o.order_id);
    expect(collected).toEqual(['ord_1', 'ord_2', 'ord_3']);
    expect(calls.length).toBe(2);
  });

  it('list passes status + limit as query params', async () => {
    const h = harness();
    h.setResponse({ orders: [] });
    const r = new CryptoOrdersResource(h.http);
    await r.list({ status: 'paid', limit: 25 });
    expect(h.calls[0]?.query).toEqual({ status: 'paid', limit: 25 });
  });

  it('V-666.BX list passes created_after / created_before through', async () => {
    const h = harness();
    h.setResponse({ orders: [] });
    const r = new CryptoOrdersResource(h.http);
    await r.list({
      created_after: '2026-05-01T00:00:00Z',
      created_before: '2026-05-12T00:00:00Z',
    });
    expect(h.calls[0]?.query).toEqual({
      created_after: '2026-05-01T00:00:00Z',
      created_before: '2026-05-12T00:00:00Z',
    });
  });

  it('get encodes order_id in the path', async () => {
    const h = harness();
    h.setResponse({ order_id: 'ord/with space' });
    const r = new CryptoOrdersResource(h.http);
    await r.get('ord/with space');
    expect(h.calls[0]?.path).toBe('/v1/billing/crypto-orders/ord%2Fwith%20space');
  });

  it('updateNote PATCHes with the customer_note body', async () => {
    const h = harness();
    h.setResponse({});
    const r = new CryptoOrdersResource(h.http);
    await r.updateNote('ord_a', { customer_note: 'PO-9' });
    expect(h.calls[0]?.method).toBe('PATCH');
    expect(h.calls[0]?.path).toBe('/v1/billing/crypto-orders/ord_a');
    expect(h.calls[0]?.body).toEqual({ customer_note: 'PO-9' });
  });

  it('cancel POSTs the cancel sub-route', async () => {
    const h = harness();
    h.setResponse({});
    const r = new CryptoOrdersResource(h.http);
    await r.cancel('ord_a');
    expect(h.calls[0]?.method).toBe('POST');
    expect(h.calls[0]?.path).toBe('/v1/billing/crypto-orders/ord_a/cancel');
  });

  it('receipt GETs the JSON receipt sub-route', async () => {
    const h = harness();
    h.setResponse({});
    const r = new CryptoOrdersResource(h.http);
    await r.receipt('ord_a');
    expect(h.calls[0]?.method).toBe('GET');
    expect(h.calls[0]?.path).toBe('/v1/billing/crypto-orders/ord_a/receipt');
  });
});
